src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultTagHelperDescriptionFactory.cs

```diff
@@ -0,0 +1,259 @@
+            if (associatedTagHelperInfos.Count == 0)
+            {
+                markdown = null;
+                return false;
+            }
+
+            var descriptionBuilder = new StringBuilder();
+            for (var i = 0; i < associatedTagHelperInfos.Count; i++)
+            {
+                var descriptionInfo = associatedTagHelperInfos[i];
+
+                if (descriptionBuilder.Length > 0)
+                {
+                    descriptionBuilder.AppendLine();
+                    descriptionBuilder.AppendLine("---");
+                }
+
+                descriptionBuilder.Append("**");
+                var tagHelperType = descriptionInfo.TagHelperTypeName;
+                var reducedTypeName = ReduceTypeName(tagHelperType);
+                descriptionBuilder.Append(reducedTypeName);
+                descriptionBuilder.AppendLine("**");
+                descriptionBuilder.AppendLine();
+
+                var documentation = descriptionInfo.Documentation;
+                if (!TryExtractSummary(documentation, out var summaryContent))
+                {
+                    continue;
+                }
+
+                var finalSummaryContent = CleanSummaryContent(summaryContent);
+                descriptionBuilder.AppendLine(finalSummaryContent);
+            }
+
+            markdown = descriptionBuilder.ToString();
+            return true;
+        }
+
+        // Internal for testing
+        internal static string CleanSummaryContent(string summaryContent)
+        {
+            // Cleans out all <see cref="..." /> and <seealso cref="..." /> elements. It's possible to
+            // have additional doc comment types in the summary but none that require cleaning. For instance
+            // if there's a <para> in the summary element when it's shown in the completion description window
+            // it'll be serialized as html (wont show).
+
+            var crefMatches = ExtractCrefRegex.Value.Matches(summaryContent).Reverse();
+            var summaryBuilder = new StringBuilder(summaryContent);
+
+            foreach (var cref in crefMatches)
+            {
+                if (cref.Success)
+                {
+                    var value = cref.Groups[2].Value;
+                    var reducedValue = ReduceCrefValue(value);
+                    reducedValue = reducedValue.Replace("{", "<").Replace("}", ">");
+                    summaryBuilder.Remove(cref.Index, cref.Length);
+                    summaryBuilder.Insert(cref.Index, $"`{reducedValue}`");
+                }
+            }
+            var lines = summaryBuilder.ToString().Split(new[] { '\n' }, StringSplitOptions.None).Select(line => line.Trim());
+            var finalSummaryContent = string.Join(Environment.NewLine, lines);
+            return finalSummaryContent;
+        }
+
+        // Internal for testing
+        internal static bool TryExtractSummary(string documentation, out string summary)
```

Comment:
Curious. Would it have made your life easier if you used Roslyn to parse these documentation?

Reply:
Yup but they need a Roslyn solution in order to parse it correctly. Something we don't have on the LanguageServer side.